homeassistant/components/braviatv/__init__.py

```diff
@@ -110,178 +112,165 @@ def __init__(
-
-        if power_status == "active":
-            self.is_on = True
-            if self._refresh_volume() and self._refresh_channels():
-                self._refresh_playing_info()
-                return
+            if not self.connected:
+                self.connected = await self.client.connect(
+                    pin=self.pin, clientid=CLIENTID_PREFIX, nickname=NICKNAME
+                )
 
-        self.is_on = False
+        self.is_on = self.connected and power_status == "active"
 
-    async def _async_update_data(self) -> None:
-        """Fetch the latest data."""
-        if self.state_lock.locked():
+        if self.is_on is False:
             return
 
-        await self.hass.async_add_executor_job(self._update_tv_data)
+        if not self.source_map:
+            await self._async_update_sources()
+        await self._async_update_volume()
+        await self._async_update_playing()
+
+    async def _async_update_sources(self) -> None:
+        """Update sources."""
+        self.source_list = []
+        self.source_map = {}
+
+        externals = await self.client.get_external_status()
+        self._sources_extend(externals, "input")
+
+        apps = await self.client.get_app_list()
+        self._sources_extend(apps, "app")
+
+        channels = await self.client.get_content_list_all("tv")
+        self._sources_extend(channels, "channel")
+
+    async def _async_update_volume(self) -> None:
+        """Update volume information."""
+        volume_info = await self.client.get_volume_info()
+        volume_level = volume_info.get("volume")
+        if volume_level is not None:
+            self.volume_level = volume_level / 100
+            self.volume_muted = volume_info.get("mute", False)
+            self.volume_target = volume_info.get("target")
+
+    async def _async_update_playing(self) -> None:
+        """Update current playing information."""
+        playing_info = await self.client.get_playing_info()
+        self.media_title = playing_info.get("title")
+        self.media_uri = playing_info.get("uri")
+        self.media_duration = playing_info.get("durationSec")
+        if program_title := playing_info.get("programTitle"):
+            self.media_title = f"{self.media_title}: {program_title}"
+        if self.media_uri:
+            source = self.source_map.get(self.media_uri, {})
+            self.source = source.get("title")
+            self.is_channel = self.media_uri[:2] == "tv"
+            if self.is_channel:
+                self.media_content_id = playing_info.get("dispNum")
+                self.media_content_type = MEDIA_TYPE_CHANNEL
+            else:
+                self.media_content_id = self.media_uri
+                self.media_content_type = None
+        else:
+            self.source = None
+            self.is_channel = False
+            self.media_content_id = None
+            self.media_content_type = None
+        if not playing_info:
+            self.media_title = "App"
```

Comment:
I wonder if we should add here:
```python
self.media_content_type = MEDIA_TYPE_APP
```

Reply:
And maybe we should change `App` to `SmartTV`? We really don't know what is running when `playing_info is None`.

Reply:
> And maybe we should change `App` to `SmartTV`? We really don't know what is running when `playing_info is None`.

It won't be a breaking change? :)